### PR TITLE
[api] enable DIAG_GET query function into commissioner

### PR DESCRIPTION
**Pull request by ZhangLe2016**
This PR adds DIAG_GET query command into the commissioner module, all owing users to query diagnostic TLV (ExtMacAddr) from a peer Thread device by setting the device's mesh local address than TLV mask bits flag.

Test:
```
> config set pskc 445f2b5ca6f2a93a55ce570a70efeecb

[done]
> start 192.168.9.2 49154
[done]
> netdiag query extaddr

Peer Address: fda1:7966:fc90:d97:0:ff:fe00:bc00
Content: {
    "ExtMacAddr": "6e56946f67040080"
}
[done]
```

### CODE REVIEW

File: include/commissioner/commissioner.hpp [hunk 1]
```diff
@@ -1147,6 +1161,36 @@ class Commissioner
...
+    /**
+     * @brief Asynchronously query diagnostic decoded data from a Thread device.
+     *
+     * This method sends a DIAG_GET.qry message to the specified Thread device,
+     * requesting the set of diagnostic data indicated by `aDiagDataFlags`.
+     * The response, or any errors encountered, will be delivered to the provided `aHandler`.
+     *
+     * @param[in, out] aHandler        A handler to process the response or any errors.
+     *                                 This handler is guaranteed to be called.
+     * @param[in]      aAddr           Mesh local address of the target Thread device.
```
Comment (wgtdkp) [inline include/commissioner/commissioner.hpp:1174]: Can the `aAddr` be a multicast address?

File: src/app/cli/interpreter.cpp [hunk 2]
```diff
@@ -279,6 +281,7 @@ const std::map<std::string, std::string> &Interpreter::mUsageMap = *new std::map
               "panid conflict <panid>"},
     {"energy", "energy scan <channel-mask> <count> <period> <scan-duration> <dst-addr>\n"
                "energy report [<dst-addr>]"},
+    {"diag", "diag query [extmac | rloc16] <dest mesh local address>"},
```
Comment (wgtdkp) [inline src/app/cli/interpreter.cpp:284]: OpenThread uses `extaddr` for Extended Mac Address, probably better to be consistent here.
```suggestion
    {"diag", "diag query [extaddr | rloc16] <dest mesh local address>"},
```

File: tests/integration/test_cli.sh [hunk 3]
```diff
@@ -54,3 +54,16 @@ test_get_commissioner_dataset()
...
+    start_daemon
+    form_network "${PSKC}"
+
+    start_commissioner "${NON_CCM_CONFIG}"
+    petition_commissioner
+    send_command_to_commissioner "diag query extmac"
+    stop_commissioner
+
+    stop_daemon
+}
```
Comment (wgtdkp) [inline tests/integration/test_cli.sh:69]: Please add a new line at the end of this file.

Comment (ZhangLe2016) [reply tests/integration/test_cli.sh]: done, thanks.

File: src/app/cli/interpreter.cpp (hunk 2 above)
Comment (ZhangLe2016) [reply src/app/cli/interpreter.cpp]: done, thanks

File: include/commissioner/commissioner.hpp (hunk 1 above)
Comment (ZhangLe2016) [reply include/commissioner/commissioner.hpp]: so far this api only supports unicast address, I update the comments here, thanks.

Comment (wgtdkp) [reply include/commissioner/commissioner.hpp]: Does the spec support multicast address?

File: include/commissioner/commissioner.hpp [hunk 4]
```diff
@@ -1147,6 +1161,36 @@ class Commissioner
      *         git repository.
      */
     static std::string GetVersion(void);
+
+    /**
+     * @brief Asynchronously query diagnostic decoded data from a Thread device.
+     *
+     * This method sends a DIAG_GET.qry message to the specified Thread device,
+     * requesting the set of diagnostic data indicated by `aDiagDataFlags`.
+     * The response, or any errors encountered, will be delivered to the provided `aHandler`.
```
Comment (wgtdkp) [inline include/commissioner/commissioner.hpp:1170]: Could you explain how the requested Diag data will be received? I know it's through the `CommissionerHandler` callbacks, but it should be documented here.

File: include/commissioner/commissioner.hpp (hunk 1 above)
Comment (wgtdkp) [inline include/commissioner/commissioner.hpp]: Does this have to be the Mesh-local address? Can link-local or OMR address work?

File: include/commissioner/commissioner.hpp [hunk 5]
```diff
@@ -1147,6 +1161,36 @@ class Commissioner
...
+     *
+     */
+    virtual void CommandDiagGetQuery(ErrorHandler aHandler, const std::string &aAddr, uint64_t aDiagDataFlags) = 0;
+
+    /**
+     * @brief Synchronously query diagnostic decoded data from a Thread device.
+     *
+     * This method sends a DIAG_GET.qry message to the specified Thread device,
+     * requesting the set of diagnostic data indicated by `aDiagDataFlags`.
+     * The method blocks until a response is received, an error occurs.
```
Comment (wgtdkp) [inline include/commissioner/commissioner.hpp:1185]: Here should clarify that the response indicates only that the query is delivered to the target device, doesn't mean the requested diag data is received.

File: src/app/cli/interpreter.cpp [hunk 6]
```diff
@@ -200,7 +202,7 @@ const std::map<std::string, Interpreter::Evaluator> &Interpreter::mEvaluatorMap
     {"announce", &Interpreter::ProcessAnnounce},   {"panid", &Interpreter::ProcessPanId},
     {"energy", &Interpreter::ProcessEnergy},       {"exit", &Interpreter::ProcessExit},
     {"quit", &Interpreter::ProcessExit},           {"help", &Interpreter::ProcessHelp},
-    {"state", &Interpreter::ProcessState},
+    {"state", &Interpreter::ProcessState},         {"diag", &Interpreter::ProcessDiag},
```
Comment (wgtdkp) [inline src/app/cli/interpreter.cpp:205]: The OpenThread `diag` command is for a different thing.
```suggestion
    {"state", &Interpreter::ProcessState},         {"netdiag", &Interpreter::ProcessNetworkDiag},
```

File: src/library/commissioner_impl.cpp [hunk 7]
```diff
@@ -621,6 +626,72 @@ void CommissionerImpl::GetPendingDataset(Handler<PendingOperationalDataset> aHan
...
+    {
+        SuccessOrExit(error = SignRequest(request));
+    }
+#endif
+
+    LOG_INFO(LOG_REGION_MESHDIAG, "sending DIAG_GET.qry");
+    if (aAddr.empty())
+    {
+        mProxyClient.SendRequest(request, onResponse, kLeaderAloc16, kDefaultMmPort);
+    }
```
Comment (wgtdkp) [inline src/library/commissioner_impl.cpp:660]: Is this desired? If true, it should be documented in the `commissioner.hpp` API

File: include/commissioner/commissioner.hpp (hunk 1 above)
Comment (ZhangLe2016) [reply include/commissioner/commissioner.hpp]: Based on spec 8.6.2, the diagnostic messages exchange between  commissioner and thread devices is implemented by proxy commands (UDP_RX and UDP_TX), and the IPv6 address should be mesh local address.

File: src/library/commissioner_impl.cpp (hunk 7 above)
Comment (ZhangLe2016) [reply src/library/commissioner_impl.cpp:660]: Yes, it is expected, and documented into the api `the leader ALOC will be set by default if it is empty.`

File: src/app/cli/interpreter.cpp (hunk 6 above)
Comment (ZhangLe2016) [reply src/app/cli/interpreter.cpp]: changed, thank.

**Review by wgtdkp — APPROVED** [review]

**Comment by ZhangLe2016** [comment]
@jwhui , could you please review this PR for me? Thanks a lot.

**Comment by ZhangLe2016** [comment]
@jwhui Could you please take a look at my code? I would be very grateful for your feedback.

**Pull request merged**